synedrion/src/tools/bitvec.rs

```diff
@@ -1,19 +1,90 @@
-use alloc::{boxed::Box, vec};
+use alloc::{boxed::Box, string::String, vec, vec::Vec};
 use core::ops::BitXorAssign;
+use digest::XofReader;
 
 use rand_core::CryptoRngCore;
 use serde::{Deserialize, Serialize};
 use serde_encoded_bytes::{Base64, SliceLike};
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub(crate) struct BitVec(#[serde(with = "SliceLike::<Base64>")] Box<[u8]>);
+#[derive(Serialize, Deserialize)]
+struct PackedBitVec {
+    bits: u32,
+    #[serde(with = "SliceLike::<Base64>")]
+    byte_vec: Box<[u8]>,
+}
+
+impl TryFrom<PackedBitVec> for BitVec {
+    type Error = String;
+    fn try_from(source: PackedBitVec) -> Result<Self, Self::Error> {
+        if source.bits.div_ceil(8) as usize > source.byte_vec.len() {
+            return Err("The declared number of bits is greater than the size of the byte string".into());
+        }
+        let bits = source
+            .bits
+            .try_into()
+            .map_err(|_| "The number of bits does not fit into `usize`")?;
+        Ok(BitVec::from_bytes_unchecked(bits, &source.byte_vec))
+    }
+}
+
+impl From<BitVec> for PackedBitVec {
+    fn from(source: BitVec) -> Self {
+        let bytes = source.0.len().div_ceil(8);
+        let mut byte_vec = vec![0u8; bytes];
+
+        // Allowing direct indexing since we set the correct vector length above.
+        #[allow(clippy::indexing_slicing)]
+        for (i, bit) in source.0.iter().enumerate() {
+            let byte_position = i / 8;
+            let bit_mask = 1 << (7 - (i % 8));
+            if *bit {
+                byte_vec[byte_position] |= bit_mask;
+            }
+        }
+
+        Self {
+            bits: source.0.len() as u32,
+            byte_vec: byte_vec.into(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(try_from = "PackedBitVec", into = "PackedBitVec")]
+pub(crate) struct BitVec(Box<[bool]>);
```

Comment:
I think I can guess the answer but why are we rolling our own bitvector type as opposed to using the [`bitvec` crate](https://crates.io/crates/bitvec)?

Reply:
I actually don't remember. We should try to use it instead.

Reply:
So I had another look at it, and, while it has all the functionality we want, the serialization is weird. It serializes `any::type_name::<Order>()` along with the actual data, and that function is not stable - in particular, its output is not guaranteed to remain the same between Rust versions. We only have about a page of code in the `bitvec` module, and it's internal and not performance-critical, so I think it's fine to leave it as is for now.